import org.jetbrains.kotlin.formver.plugin.*

@Pure
@DumpExpEmbeddings
fun <!VIPER_TEXT!>intLit<!>(x: Int, y: Int) : Int = 3

@Pure
@DumpExpEmbeddings
fun <!VIPER_TEXT!>boolLit<!>(x: Int, y: Int) : Boolean = true